### PR TITLE
4.x: TakeLast() improve structure and recursive scheduling

### PR DESCRIPTION
**Pull request by akarnokd**
- Inline the relevant parent fields into the sinks
- Use a more efficient recursive scheduling disposable-tracking scheme
- Delegate to base.Run where applicable.

### CODE REVIEW

File: Rx.NET/Source/src/System.Reactive/Linq/Observable/TakeLast.cs [hunk 1]
```diff
@@ -123,37 +131,36 @@ public Time(IObservable<TSource> source, TimeSpan duration, IScheduler scheduler
...
+                    _loopScheduler = parent._loopScheduler;
                     _queue = new Queue<System.Reactive.TimeInterval<TSource>>();
                 }
 
-                private IDisposable _sourceDisposable;
                 private IDisposable _loopDisposable;
                 private IStopwatch _watch;
 
-                public void Run()
+                public void Run(IObservable<TSource> source, IScheduler scheduler)
```
Comment (akarnokd) [inline Rx.NET/Source/src/System.Reactive/Linq/Observable/TakeLast.cs]: `scheduler` is only needed to get the `StartStopwatch` before the operator executes.

File: Rx.NET/Source/src/System.Reactive/Linq/Observable/TakeLast.cs [hunk 2]
```diff
@@ -56,32 +56,40 @@ protected override void Dispose(bool disposing)
...
+                    }
                 }
 
-                private void LoopRec(Action<_> recurse)
+                private IDisposable LoopRec(IScheduler scheduler)
                 {
                     if (_queue.Count > 0)
                     {
                         ForwardOnNext(_queue.Dequeue());
-                        recurse(this);
```
Comment (danielcweber) [inline Rx.NET/Source/src/System.Reactive/Linq/Observable/TakeLast.cs]: Can you elaborate more on the issue with this?

Comment (akarnokd) [reply Rx.NET/Source/src/System.Reactive/Linq/Observable/TakeLast.cs]: It is the very same situation with `Range` and `ToObservable`: the Action-based recursive scheduling is more expensive.

https://github.com/dotnet/reactive/pull/684/files#diff-76edecc51f6af371eae34be29849253aR62

Comment (danielcweber) [reply Rx.NET/Source/src/System.Reactive/Linq/Observable/TakeLast.cs]: Any possibility to optimize the action based recursive scheduling? The syntax is much less noisy.

Comment (akarnokd) [reply Rx.NET/Source/src/System.Reactive/Linq/Observable/TakeLast.cs]: I don't see how. It has to maintain some state across schedules. There was an optimization for it in #617, particularly the `InvokeRec1State` which holds a `CompositeDisposable` and creates a `SingleAssignmentDisposable` for each schedule invocation.

**Comment by akarnokd** [comment]
Is there something else needed for this PR?

File: Rx.NET/Source/src/System.Reactive/Linq/Observable/TakeLast.cs (hunk 2 above)
Comment (onovotny) [reply Rx.NET/Source/src/System.Reactive/Linq/Observable/TakeLast.cs]: Would local functions help? You can capture the state without needing a delegate.

Comment (akarnokd) [reply Rx.NET/Source/src/System.Reactive/Linq/Observable/TakeLast.cs]: I think that still makes an allocation just like a capturing delegate would upon each invocation.

Comment (onovotny) [reply Rx.NET/Source/src/System.Reactive/Linq/Observable/TakeLast.cs]: Yes, it does still have a state variable for the capture

Comment (danielcweber) [reply Rx.NET/Source/src/System.Reactive/Linq/Observable/TakeLast.cs]: I think the issue here is, why does the action based scheduling have to keep track of all the schedulings in a `CompositeDisposable`. Seems odd, at least for tail recursion.

**Pull request merged**